Visualization serialization

jeffbaumes opened the pull request:
This is an initial serialization prototype. `getSerializationFormats()` provides a list of serialization types, which may be sent to `serialize()`.

The examples page will produce buttons for downloading each serialization type supported by the visualization. Currently only works with the scatterplot example.

ronichoudhury commented on src/candela/components/ScatterPlot/index.js:
```diff
@@ -85,6 +85,19 @@ export default class ScatterPlot extends VisComponent {
+    if (!this.chart) {
+      this.render();
+    }
+    return this.chart.then((vobj) => {
```
I'd say get rid of the parentheses around `vobj` (unless we decided on a style that keeps them for single-argument arrow functions).

ronichoudhury commented on src/candela/components/ScatterPlot/index.js:
```diff
@@ -85,6 +85,19 @@ export default class ScatterPlot extends VisComponent {
+
+  serialize (format) {
+    if (!this.chart) {
+      this.render();
```
I don't think serializing should cause a render. In this case I'd prefer to return a fulfilled promise that delivers `undefined` or `null` or a blank image.

ronichoudhury commented:
Very cool!

jeffbaumes commented:
@ronichoudhury PTAL. I've removed parens and changed serialize to return a rejected promise, which is tested in the example index.js - I ensure that the promise is rejected if render has not been called.

ronichoudhury commented:
Rejected promise! Great idea :tada:

ronichoudhury commented:
One thing I'm wondering about: since we're using a getter to store the options spec for a `VisComponent`, should we do something similar for serialization formats? It seems like they should either both be getters, or both be methods. Thoughts?

(Aside from this discussion, the code looks pretty good, FYI.)

jeffbaumes commented:
Good point about getters. Instead adding another field to `spec`, I propose we get rid of `spec` entirely and have:

```
static get options()

static get container()

get serializationFormats()
```

`options` and `container` need to be static since you need them to initialize the visualization, while `serializationFormats` does not need to be static, so probably shouldn't be. Also the formats could in theory vary based on how the vis is initialized.

ronichoudhury commented:
I like the options/container/serializationFormats getters proposal.

jeffbaumes commented:
This is ready for final review. I think splitting spec into options/container should go in another PR.

ronichoudhury commented:
The `VegaChart` mixin looks awesome!

If we can just fix the problem with the UI buttons spoiling the image tests and Travis is otherwise happy, then LGTM :smiley_cat:

jeffbaumes commented:
I made Travis happy so I'm merging due to @ronichoudhury's provisional LGTM.

ronichoudhury commented:
:fireworks:
